warehouse/postgres/postgres.go

```diff
@@ -0,0 +1,520 @@
+	object, err := warehouseutils.GetObjectName(pg.Warehouse.Destination.Config, objectLocation)
+	if err != nil {
+		logger.Errorf("PG: Error in converting object location to object key for table:%s: %s,%v", tableName, objectLocation, err)
+		return "", err
+	}
+	dirName := "/rudder-warehouse-load-uploads-tmp/"
+	tmpDirPath, err := misc.CreateTMPDIR()
+	if err != nil {
+		logger.Errorf("PG: Error in creating tmp directory for downloading load file for table:%s: %s, %v", tableName, objectLocation, err)
+		return "", err
+	}
+	ObjectPath := tmpDirPath + dirName + fmt.Sprintf(`%s_%s_%d/`, pg.Warehouse.Destination.DestinationDefinition.Name, pg.Warehouse.Destination.ID, time.Now().Unix()) + object
+	err = os.MkdirAll(filepath.Dir(ObjectPath), os.ModePerm)
+	if err != nil {
+		logger.Errorf("PG: Error in making tmp directory for downloading load file for table:%s: %s, %s %v", tableName, objectLocation, err)
+		return "", err
+	}
+	objectFile, err := os.Create(ObjectPath)
+	if err != nil {
+		logger.Errorf("PG: Error in creating file in tmp directory for downloading load file for table:%s: %s, %v", tableName, objectLocation, err)
+		return "", err
+	}
+	downloader, err := filemanager.New(&filemanager.SettingsT{
+		Provider: warehouseutils.ObjectStorageType(pg.Warehouse.Destination.DestinationDefinition.Name, pg.Warehouse.Destination.Config),
+		Config:   pg.Warehouse.Destination.Config.(map[string]interface{}),
+	})
+	err = downloader.Download(objectFile, object)
+	if err != nil {
+		logger.Errorf("PG: Error in downloading file in tmp directory for downloading load file for table:%s: %s, %v", tableName, objectLocation, err)
+		return "", err
+	}
+	fileName := objectFile.Name()
+	if err = objectFile.Close(); err != nil {
+		logger.Errorf("PG: Error in closing downloaded file in tmp directory for downloading load file for table:%s: %s, %v", tableName, objectLocation, err)
+		return "", err
+	}
+	return fileName, nil
+}
+
+func (pg *HandleT) loadTable(tableName string, columnMap map[string]string) (err error) {
+	status, err := warehouseutils.GetTableUploadStatus(pg.Upload.ID, tableName, pg.DbHandle)
+	if status == warehouseutils.ExportedDataState {
+		logger.Infof("PG: Skipping load for table:%s as it has been successfully loaded earlier", tableName)
+		return
+	}
+	if !warehouseutils.HasLoadFiles(pg.DbHandle, pg.Warehouse.Source.ID, pg.Warehouse.Destination.ID, tableName, pg.Upload.StartLoadFileID, pg.Upload.EndLoadFileID) {
+		warehouseutils.SetTableUploadStatus(warehouseutils.ExportedDataState, pg.Upload.ID, tableName, pg.DbHandle)
+		return
+	}
+	logger.Infof("PG: Starting load for table:%s", tableName)
+	warehouseutils.SetTableUploadStatus(warehouseutils.ExecutingState, pg.Upload.ID, tableName, pg.DbHandle)
+	timer := warehouseutils.DestStat(stats.TimerType, "single_table_upload_time", pg.Warehouse.Destination.ID)
+	timer.Start()
+	defer timer.End()
+
+	dbHandle, err := connect(pg.getConnectionCredentials(optionalCredsT{schemaName: pg.Namespace}))
```

Comment:
did this in case of snowflake because using same db handle concurrently was throwing error. I think you can use pg.Db handle here instead of creating one every time. Do test it with a bigger number for maxParallelLoads